egress gw exclude internal traffic

kkourt opened the pull request:
This PR aims to do fix an issue where egress gw policy is incorrectly applied to internal cluster traffic, using a patch originally from @anfernee. In addition to that, it refactors the function `snat_v4_needed` so that it is easier to follow. Since this is brittle code, the refactoring is split into small patches so that they are easy to review.


Fixes: #16147

```release-note
fix incorrect application of egress gateway policy to internal cluster traffic.
require  a 5.2 kernel or later for the egress gateway policy feature.
```

jibi committed:
daemon: require kernel >= 5.2 for egress gateway

After datapath modifications in this PR, 4.19 kernel is unable to verify
bpf_lxc:

>  => Loading bpf_lxc.c:from-container...
>  ...
>
> Prog section '2/7' rejected: Argument list too long (7)!
>  - Type:         3
>  - Attach Type:  0
>  - Instructions: 4130 (34 over limit)
>  - License:      GPL

Introduce a requirement for egress gateway to run on 5.2 or later
kernels that support a larger number of instructions.

In addition to the daemon, modify testing and build files accordingly.

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>
Signed-off-by: Kornilios Kourtis <kornilios@isovalent.com>

kkourt committed:
k8sT/Egress: restructure test

Currently, we print the policy map after the policy is removed. This
means that the map will always be empty.

test used go's defer to remove the egress gateway policy. We modify this
to use an AfterAll block for removing this policy, which in turn, allow
us to use an AfterFailed block to print the maps.

Fixes: 25f32bca224e5e80b3148a151acf718a404b11ac

Signed-off-by: Kornilios Kourtis <kornilios@isovalent.com>

kkourt committed:
k8sT/Egress: test exclusion of internal traffic

Add an additional CiliumEgressNATPolicy that will match every packet and
forward it into 1.1.1.1. This allows us to check that internal traffic
will not be checked using the egress gw policy.

This was originally from:
https://github.com/cilium/cilium/pull/16447/commits/c9485e45dae747f4633f49a4e689b6d543cf6b61

Signed-off-by: Yongkun Gui <ygui@google.com>
Signed-off-by: Kornilios Kourtis <kornilios@isovalent.com>

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -797,6 +797,21 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx,
+		 * in cluster. In this case, we should skip egress gateway. If destination
+		 * is either remote node or host node, also skip egress gateway.
+		 */
+		if (tunnel_endpoint != 0 || *dst_id == REMOTE_NODE_ID || *dst_id == HOST_ID)
```
Isn't the `tunnel_endpoint` check only required if `IPV4_SNAT_EXCLUSION_DST_CIDR` is undefined?

kkourt replied:
I'm not sure 🤔  @brb?

brb replied:
Yeah, if the `IPV4_SNAT_EXCLUSION_DST_CIDR` is undefined, we can remove the check, as they both check for the same - whether the dst is a pod.

pchaigno replied:
Up to you @kkourt. It will reduce complexity a little bit when `IPV4_SNAT_EXCLUSION_DST_CIDR` is defined but at the cost of slightly harder to read code I guess. I don't have a preference of what to do.

kkourt replied:
How about we continue this discussion in: https://github.com/cilium/cilium/pull/17688#discussion_r735408426, so that we do not block this PR?

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1207,6 +1188,22 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
+			*from_endpoint = true;
+			return true;
+		}
+	}
```
Instead of checking if the destination is a remote node here, couldn't we extend the previous check (`info->sec_label == REMOTE_NODE_ID` on line 1188) to also run if `!defined(TUNNEL_MODE) && defined(ENABLE_EGRESS_GATEWAY)`? Then, we wouldn't need this code here (lines 1191--1206) and we could run the next egressgw map lookup regardless of whether `ep` is NULL or not. It seems it would simplify the code a bit.

The indentation for this block also seems broken.

kkourt replied:
> couldn't we extend the previous check 

The previous check will `return false`. Do you mean to `return false` if `info->sec_label == REMOTE_NODE_ID` when `ENABLE_EGRESS_GATEWAY` is defined? Would this be correct? I am not sure.

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1189,20 +1189,20 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
+					*from_endpoint = true;
+					return true;
+				}
+			}
```
Please merge `bpf: snat_v4_needed: fix indentation` in previous patch.

kkourt replied:
Makes sense will do.

brb replied:
Let's do once all reviews are done :-)

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1157,6 +1157,20 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
 		return false;
 #endif
 
+#ifdef ENABLE_IP_MASQ_AGENT
```
Nit: Commit title has a small typo: `bpf: snat_v4_needed: move ip-masq-agent further ip` (`ip` -> `up`).

brb replied:
I think it's been moved too much further up. LPM trie lookup can be expensive. We should do that only if we know that the src is an endpoint.

kkourt replied:
OK, dropping this patch.

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1157,71 +1158,72 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
+	if (ep && ep->flags & ENDPOINT_F_HOST)
+		return false;
+
+	*from_endpoint = true;
```
So `from_endpoint` used to be true only if the source endpoint was local to the node, now it will also be true if the source endpoint is remote. AFAICS, that only happens for the egress gateway case. The only consumer I found of this variable is `snat_v4_can_skip`, but I'm not able to tell if this change could cause issues there. Maybe @brb can have a second look.

(@brb `snat_v4_can_skip` could really use a couple comments to clarify the intent of the different conditions.)

kkourt replied:
Ah good catch.

I guess maintaining the original semantics would mean doing:
```
if (ep)
   *from_endpoint = true;
```

brb replied:
With the introduction of egress GW `from_endpoint` has changed it meaning - previously it meant a local endpoint, now it means both - local endpoint and a remote endpoint which is a target of egress GW policy.

> (@brb snat_v4_can_skip could really use a couple comments to clarify the intent of the different conditions.)

:+1:

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1167,17 +1181,6 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
-		memcpy(pfx.lpm.data, &ip4->daddr, sizeof(pfx.addr));
-		if (map_lookup_elem(&IP_MASQ_AGENT_IPV4, &pfx))
-			return false;
-#endif
```
This commit doesn't just optimize but also changes the logic. We used to only lookup the ipmasq map if (1) the source was a local endpoint and (2) the destination was present in the ipcache. This logic was introduced by commit 0a6160f6102 ("datapath: Add ip-masq-agent implementation"). The commit description makes me think that (1) might have been intentional. (2) seems stranger but I suspect @brb had a good reason to add the map lookup in this exact location.

kkourt replied:
I did this change together with @brb and said it was fine, but I guess it is worth another look.

brb replied:
Could you elaborate why the logic is changed in this commit? (1) and (2) premises are still there?

Re (2): scratching my head, but I don't see any reason why the ip-masq-agent check should depend on (2).

brb replied:
Oh, GH was showing this comment already for the previous commit which confused me :F

pchaigno replied:
Ok. So we don't care about (2) but we do for (1) for performance reasons if I'm reading https://github.com/cilium/cilium/pull/17639/commits/64357b35302cb7536f7833b405598216aed083ce#r735072541 correctly.

Also, now that I think about it, I suspect this commit will have a negative complexity impact. In general (though not always true), the further down in the control flow graph you can move a piece of code, the better it is for complexity.

brb replied:
>  So we don't care about (2) but we do for (1) for performance reasons

Yep.

kkourt replied:
Since this is more complicated that I  originally thought, I'll drop the patch for now.

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1211,19 +1209,24 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
 
-		tuple.nexthdr = ip4->protocol;
-		tuple.daddr = ip4->daddr;
-		tuple.saddr = ip4->saddr;
```
I'm not sure I see the benefit of `bpf: snat_v4_needed: refactor reply check`. If anything, the incorrect indentation (w.r.t. to the conditions) feels misleading to me. If the goal is to have the value of `is_reply` at hand when reading the consumer code, maybe moving the initialization of this variable here would be enough?

kkourt replied:
For me it's easier to understand what is going on if variables are declared exactly when they are used. Because by looking at this:
```
+               {
+                       bool is_reply = false;
+                       struct ipv4_ct_tuple tuple = {
+                               .nexthdr = ip4->protocol,
+                               .daddr = ip4->daddr,
+                               .saddr = ip4->saddr,
+                       };
+
+                       /* The packet is a reply, which means that outside
+                        * has initiated the connection, so no need to SNAT
+                        * the reply.
+                        */
+                       if (!ct_is_reply4(get_ct_map4(&tuple), ctx,
+                                         ETH_HLEN + ipv4_hdrlen(ip4),
+                                         &tuple, &is_reply) &&
+                           is_reply)
+                               return false;
+               }
```

instead of this:


```
-               tuple.nexthdr = ip4->protocol;
-               tuple.daddr = ip4->daddr;
-               tuple.saddr = ip4->saddr;
-
-               /* The packet is a reply, which means that outside
-                * has initiated the connection, so no need to SNAT
-                * the reply.
-                */
-               if (!ct_is_reply4(get_ct_map4(&tuple), ctx,
-                                 ETH_HLEN + ipv4_hdrlen(ip4),
-                                 &tuple, &is_reply) &&
-                   is_reply)
-                       return false;
```

The following questions have obvious answers:
 * are there fields in the tuple that are initialized elsewhere?
 * is the tuple and is_reply used somewhere after this function?

Might be just me though :)

pchaigno replied:
I had not though about that second point (clarifies scope of variables); that's a good point. It's really a nit and we don't have a consistent rule today, so I'm fine either way :-)

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1231,22 +1233,6 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
-			return true;
-		}
-	}
-#endif
```
Commit `bpf: snat_v4_needed: single egress gw policy check` can be merged into the commit it fixes if you move the refactoring logic to the beginning of the PR's commit history. That way reviewers wouldn't waste time commenting on the first commits...

kkourt replied:
OK, if people think that the logic is correct, I think the best approach is what you said: fold: the original fix, the indentation fix, and this one into one commit. and the refactoring fixes we can put on top (or remove).

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1175,9 +1175,7 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
 	*from_endpoint = true;
-
-	info = ipcache_lookup4(&IPCACHE_MAP, ip4->daddr,
-			       V4_CACHE_KEY_LEN);
```
I'm guessing commit `bpf: snat_v4_needed: misc refactor cleanups` was intended to contain more misc. changes as you implemented but didn't end up containing any others? This simple newline could be removed in the commit where you removed the indentation; that wouldn't make it harder to review.

kkourt replied:
Yeah, not sure what the best approach here. I could also fold this one as well. @brb said that it would be easier to review if the modifications were small and obvious. I probably took that too far :)

pchaigno replied:
My rationale here is: (1) this change is only possible if a previous commit (the one fixing indentation) exists and (2) it is trivial. Thus, I'd merge both commits (and then would merge in the first commit they both fix). Another way I'd approach this is: style fixes don't merit their own commit unless the code they fix is already in master.

pchaigno commented on test/k8sT/Egress.go:
```diff
@@ -106,7 +106,7 @@ var _ = SkipDescribeIf(func() bool {
 
 	AfterFailed(func() {
 		// Especially check if there are duplicated address allocated on cilium_host
-		kubectl.CiliumReport("ip addr", "cilium bpf egress list", "cilium bpf nat list")
```
Nit: Seems there might be a small typo somewhere in clause `which resulted in commands that were executed at failure to diagnose issues to with the policy removed.` in the commit description (commit `k8sT/Egress: restructure test`). I'm not able to parse it.

kkourt replied:
thanks, will fix.

kkourt replied:
wow, the message was really bad. Fixed it.

pchaigno commented on test/k8sT/Egress.go:
```diff
@@ -179,22 +179,33 @@ var _ = SkipDescribeIf(func() bool {
+					kubectl.WaitForEgressPolicyEntry(k8s2IP, outsideIP)
+				})
+				AfterAll(func() {
+					defer kubectl.Delete(policyYAML)
```
```suggestion
					kubectl.Delete(policyYAML)
```

pchaigno commented on Documentation/gettingstarted/egress-gateway.rst:
```diff
@@ -12,6 +12,10 @@ Egress Gateway (beta)
 
 .. include:: ../beta.rst
 
+.. note::
```
In commit `daemon: require kernel >= 5.4 for egress gateway`, seems a part of the first sentence is missing:

> After the modifications in recent , 4.19 kernel is unable to verify bpf_lxc:

pchaigno commented on Documentation/gettingstarted/egress-gateway.rst:
```diff
@@ -12,6 +12,10 @@ Egress Gateway (beta)
 
+.. note::
+
+   Egress Gateway requires a 5.4 or more recent kernel.
```
The 1M instructions limit was introduced in Linux v5.2, with commit https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=c04c0d2b968ac45d6ef020316808ef6c82325a82.
```suggestion
   Egress Gateway requires a 5.2 or more recent kernel.
```

kkourt replied:
@jibi any reason this was `5.4` in the original commit?

jibi replied:
Just me mistaking versions

pchaigno commented on Documentation/operations/system_requirements.rst:
```diff
@@ -186,6 +186,7 @@ Full support for :ref:`session-affinity`    >= 5.7
 BPF-based proxy redirection                 >= 5.7
 BPF-based host routing                      >= 5.10
 Socket-level LB bypass in pod netns         >= 5.7
+:ref:`egress-gateway`                       >= 5.4
```
```suggestion
:ref:`egress-gateway`                       >= 5.2
```

pchaigno commented on bpf/Makefile:
```diff
@@ -169,7 +169,9 @@ HOST_OPTIONS = $(LXC_OPTIONS) \
-ifneq ("$(KERNEL)","49")
+
+# Egress Gateway should only be enabled for >= 5.4 kernels
+ifneq (,$(filter $(KERNEL),"54" "netnext"))
```
Oh, wow! We can do that? :exploding_head:

pchaigno commented on daemon/cmd/daemon.go:
```diff
@@ -663,6 +664,11 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 	}
+	if option.Config.EnableEgressGateway {
+		if !probes.NewProbeManager().GetMisc().HaveLargeInsnLimit {
+			return nil, nil, fmt.Errorf("egress gateway needs kernel 5.4 or newer")
```
```suggestion
			return nil, nil, fmt.Errorf("egress gateway needs kernel 5.2 or newer")
```

pchaigno commented on test/helpers/utils.go:
```diff
@@ -515,7 +515,7 @@ func DoesNotRunOn419Kernel() bool {
 
-// RunsOn419OrLaterKernel checks whether a test case is running on the bpf-next,
+// RunsOn419OrLaterKernel checks whether a test case is running on 4.19 or later kernel
 // 4.19.x (x > 57), or 5.4 kernels.
```
```suggestion
// RunsOn419OrLaterKernel checks whether a test case is running on 4.19 or
// later kernel.
```

pchaigno commented on test/k8sT/Egress.go:
```diff
@@ -25,7 +25,7 @@ import (
 
 var _ = SkipDescribeIf(func() bool {
-	return helpers.RunsOnEKS() || helpers.RunsOnGKE() || helpers.DoesNotRunWithKubeProxyReplacement() || helpers.DoesNotExistNodeWithoutCilium()
+	return helpers.RunsOnEKS() || helpers.RunsOnGKE() || helpers.DoesNotRunWithKubeProxyReplacement() || helpers.DoesNotExistNodeWithoutCilium() || helpers.DoesNotRunOn54Kernel()
```
```suggestion
	return helpers.RunsOnEKS() || helpers.RunsOnGKE() || helpers.DoesNotRunWithKubeProxyReplacement() || helpers.DoesNotExistNodeWithoutCilium() || helpers.DoesNotRunOn54OrLaterKernel()
```

pchaigno reviewed (changes requested):
The code changes look good to me overall and it's nice to see a bit of refactoring to the SNAT logic. I think some of the changes need a careful review from @brb as they change the logic a bit besides the intended refactoring/optimization.

The commit history however makes it painful to review. It seems a lot of care was taken to avoid changing the first commit—I'm not sure why. As a result, comments on the first commit end up being addressed by later commits and I had to review the same piece of code several times. I've tried to leave comments on how that can be improved.

(If we need this merged quickly, I think several pieces of refactoring could be moved to another pull request as the egress gateway fix does not depend on them. In general, I'd advise for such smaller pull requests when possible anyway. Easier to review, faster to merge, easier to backport.)

brb commented on bpf/lib/nodeport.h:
```diff
@@ -1207,6 +1188,22 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
 			if (info->sec_label == REMOTE_NODE_ID)
 				return false;
 #endif
+#if defined(ENABLE_EGRESS_GATEWAY) && !defined(IS_BPF_OVERLAY)
```
Having comment why `!defined(IS_BPF_OVERLAY)` would be useful.

kkourt replied:
As @brb pointed offline, there is a check some lines above for this, so I'll remove it

brb commented on bpf/lib/nodeport.h:
```diff
@@ -1175,42 +1155,67 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
+	/* If this is a localhost endpoint, no SNAT is needed. */
+	if (ep && ep->flags & ENDPOINT_F_HOST)
+		return false;
+
```
Maybe we should add comment here saying that "at this point we know, that the src is either a local endpoint or a forwarded packet which might be the subject to egress GW SNAT".

kkourt replied:
Ended up dropping this patch since @pchaigno pointed out some complications that I missed. We can revisit refactoring the function in another PR.

jibi commented on bpf/lib/nodeport.h:
```diff
@@ -1207,6 +1188,25 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
+			 *  1. destination is a not a remote node
+			 *  2. there is no local endpoint for source address
+			 */
+			if (info->sec_label != REMOTE_NODE_ID || !ep) {
```
I'm not sure I'm following this check. Policies apply only to traffic originating from an endpoint and destined to the outside of the cluster, so why not explicitly check `if (ep && info->sec_label == WORLD_ID)`?

To my understanding with this check we may end up doing some lookups in the policy map when we should not (resulting in internal cluster traffic being SNAT'ed):

* with `info->sec_label != REMOTE_NODE_ID` we are making sure traffic is not destined to a remote node in the cluster, but it could still be destined to a remote endpoint (and so should not be SNAT'ed)
* with `!ep` we are making sure traffic is not originating from an endpoint, but why? Egress policies should always match traffic originating from an endpoint, so we should never find a policy in the map if `ep` is `NULL`

kkourt replied:
Hm... I also just realized that `!ep` is never true in this case, so the patch does not make any sense as it is :(

kkourt replied:
> so why not explicitly check if (ep && info->sec_label == WORLD_ID)?

So the first thing I would like to resolve is whether we only need to apply egress gw policy iff `ep != NULL`. (above comment seems to imply that this is the case).

ep is:
```
ep = __lookup_ip4_endpoint(ip4->saddr);
```

And so according to my understanding we will perform the SNAT at the gateway node where we might not have an endpoint for the source address. So I think we should apply the egress policy check both when `ep` and `!ep`. Let me try another version...

kkourt replied:
Pushed a new version with what, according to my understanding, we need to do. Please let me know what you think.

kkourt replied:
After discussing this offline, we settled on using ` info->sec_label == WORLD_ID` as a test.

kkourt replied:
We discussed this with @pchaigno offline, and decided to not use WORLD_ID because, as Paul pointed out, egress traffic might not have a WORLD_ID because it matches CIDR policies.

pchaigno reviewed (changes requested):
Were all changes merged by mistake into a single commit? It now seems all refactoring changes are in the first commit, with the fix itself. The commit description doesn't reflect that (or even describe the fix correctly) and it's obviously harder to review. Happy to help split it correctly if needed.

kkourt committed:
bpf: snat_v4_needed: exit early if ep is local

This is a preparation patch (1/2) for subsequent changes. The patch
intends to maintain the code behavior.

Signed-off-by: Kornilios Kourtis <kornilios@isovalent.com>

kkourt committed:
bpf: snat_v4_needed: move declerations closer to use

This is a preparation patch for (2/2) subsequent changes. The patch
intends to maintain the code behavior.

Signed-off-by: Kornilios Kourtis <kornilios@isovalent.com>

kkourt committed:
datapath: Fix egress gw to exclude in-cluster traffic

datapath code checks the egress gateway map to determine whether the packet
should be forwarded to the egress gateway and whether the packet should
be SNAT-ed with an appropriate egress IP.

In-cluster traffic should be, however, excluded from these checks.
In-cluster traffic includes both pod-to-pod and pod-to-node. The purpose
of this patch is to distinguish between cluster-local versus
clluster-egress traffic, and apply the egress map check only in the
latter case.

There are two check points:
 - in bpf_lxc we check whether the packet should be forwarded to
 the egress gateway (currently, via a tunnel)
 - in bpf_host (in the gateway node) we check whether the packet should
 be SNATed

This patch adds the following tests to predicate checking the egress gw
map.

In bpf_lxc, we test whether:
 - there is tunnel_endpoint in ipcache, which means the destination is a
 remote endpoint
 - the destination id is REMOTE_NODE_ID or HOST_ID, which means that the
 destination is the host or a remote node
 - the destination matches a local endpoint
and in all above cases we skip checking the egress policy map.

In bpf_host, we only perform the egress map check if:
 - the source address does not match a local endpoint
 - the remote address is not a remote node

One of the things that we considered was using WORLD_ID to distinguish
egress traffic. However, as Paul pointed out, egress traffic might
not have a WORLD_ID because it might match CIDR policies (toCIDR rules).

Fix issue:16147

Signed-off-by: Yongkun Gui <ygui@google.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Kornilios Kourtis <kornilios@isovalent.com>

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1208,8 +1189,31 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
+		 *  - is not destined for a remote node
+		 */
+		if (!ep || info->sec_label != REMOTE_NODE_ID) {
+			struct egress_info *einfo;
```
The code and comment are consistent with the commit description, but I think the comment could use a bit more detail. Maybe something like:
```
/* Check egress gateway policy only for traffic which matches one of the
 * following conditions.
 *  - Not from a local endpoint (inc. local host): that tells us the
 *    traffic was redirected by an egress gateway policy to this node to
 *    be masqueraded.
 *  - Not destined for a remote node: that tells us the traffic is leaving
 *    the cluster. Inter-node traffic to remote pods would either leave
 *    through the tunnel or match the above IPV4_SNAT_EXCLUSION_DST_CIDR
 *    check.
 */
```

(Worth noting this assumes `IPV4_SNAT_EXCLUSION_DST_CIDR` will always be defined when both native routing and BPF masquerading are enabled, which I think is true.)

kkourt replied:
added, thanks!

pchaigno reviewed (approved):
One minor comment on a code comment below, but the logic for the fix & refactoring looks good to me! Thanks for the cleaner commit history as well! :pray:

qmonnet commented on test/k8sT/manifests/egress-nat-policy.yaml:
```diff
@@ -13,3 +13,21 @@ spec:
+  # It shouldn't affect in-cluster traffic.
+  destinationCIDRs:
+  - 0.0.0.0/0
+  egressSourceIP: 1.1.1.1 # It's a black hole
```
Probably doesn't matter much, but I think using 1.1.1.1 as a black hole is an abuse of the standards. Traffic probably get discarded anyway, but there _is_ something at that address now - Cloudflare's DNS and some web page at least. (But then I don't have a better suggestion.)

jibi replied:
Maybe we can use a documentation address for example from the `192.0.2.0/24` block? (See https://datatracker.ietf.org/doc/html/rfc5737#section-3)

qmonnet replied:
Sound like a good idea to me

kkourt replied:
Thanks, used 192.0.2.13

qmonnet reviewed (approved):
Looks good to me

christarazi reviewed (commented):
Came across this PR and noticed the first two commits don't have a descriptive title describing what is changing and why, which made it harder to understand why it is needed. They only state "refactor x/2". If it's not too much of a hassle, could we add a bit more to it?

brb commented on bpf/lib/nodeport.h:
```diff
@@ -1116,11 +1117,11 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
 	 * cluster.
 	 */
 	if (1) {
```
While we are here - any reason for `if (1)`?

kkourt replied:
I don't think so, seems to me it's just a way of defining a scope.

brb commented on bpf/lib/nodeport.h:
```diff
@@ -1176,55 +1177,56 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
+	 * no SNAT is needed.
+	 */
+	if (!ep || (ep->flags & ENDPOINT_F_HOST))
+		return false;
```
Thinking out loud: this branch is very unlikely to happen, as for majority of cases it's going to be from pod or from host traffic. Also, the check against `IPV4_MASQUERADE` above should catch the localhost endpoint case for majority of cases. Maybe it's a good use case for `unlikely()`?

kkourt replied:
I can add it. I've seen it being used elsewhere in the datapath, but have no idea if it generally helps or not.

(I generally avoid it for x86, since I've never been able to see any perf improvement when using it.)

pchaigno replied:
I've seen noticeable improvements on DPDK code. Never checked on BPF. Compiler definitely takes it into account since we've since it fix a complexity issue once (just because code was reordered).

brb replied:
Probably @borkmann could comment on this too.

borkmann replied:
I would leave this case probably as is rather than marking as unlikely() given afaiu this could be hit under multi-dev case.

brb reviewed (approved):
Thanks! Also, :+1: for the detailed commit msg. A few non-blocking comments. Also, out of curiosity - have you checked what is a complexity increase after the refactoring in the bpf_host?

kkourt committed:
daemon: require remote-node-identity for egress gw

Egress gateway datapath code depends on REMOTE_NODE_ID to distinguish
between cluster-local and cluster-egress traffic.

Currently, egress gateway depends on BPF masquerade, which also requires
remote-node-identity. However, this might change in the future so we add
a check for the above.

Signed-off-by: Kornilios Kourtis <kornilios@isovalent.com>
